### PR TITLE
Ece mixtures

### PR DESCRIPTION
**Pull request by ericward-noaa**
NWFSC / SWFSC assessors requested adding another family of models, for shoaling spp (e.g. Thorson et al. 2011 -- https://doi.org/10.1139/f2011-086). 

Previous models had treated positive part as a mixture (categorical sampling in JAGS) -- this approach uses the marginalization approach.

Implemented here as "gamma_mix" or "delta_gamma_mix" -- but lognormal equivalents could be added too

### CODE REVIEW

File: R/fit.R [hunk 1]
```diff
@@ -1100,6 +1104,10 @@ sdmTMB <- function(
   tmb_map$b_j <- NULL
   if (delta) tmb_map$b_j2 <- NULL
   if (family$family[[1]] == "tweedie") tmb_map$thetaf <- NULL
+  if (family$family[[1]] %in% c("gamma_mix", "delta_gamma_mix")) {
```
Comment (seananderson) [inline R/fit.R:1107]: Very cool! This line looks off though. `family[[1]]` will always be `'binomial'` in a delta model. This part is valid:

```r
family$family[[1]] %in% "gamma_mix"
```

but the delta option would have to ID'ed after first checking if `delta` is true and then if `family$family[[2]] == "gamma_mix"`.

```r
if (delta) {
    if (family$family[[2]] %in% "gamma_mix")
}
```

Is my guess without testing.

Comment (ericward-noaa) [reply R/fit.R]: Ah -- yes, good catch. Fixed logic in this commit

https://github.com/pbs-assess/sdmTMB/commit/633efc00aa380f38e6e164c8ecadb42b27c92362

**Comment by ericward-noaa** [comment]
Tested logic for both non-delta / delta models, and while I was doing this also added the lognormal_mix and delta_lognormal_mix distributions. FWIW I think the Gamma mixture has been preferred in the past and seems to fit better

File: R/fit.R [hunk 2]
```diff
@@ -1100,6 +1106,18 @@ sdmTMB <- function(
   tmb_map$b_j <- NULL
   if (delta) tmb_map$b_j2 <- NULL
   if (family$family[[1]] == "tweedie") tmb_map$thetaf <- NULL
+  if (family$family[[1]] %in% c("gamma_mix", "lognormal_mix")) {
+    tmb_map$log_ratio_mix <- NULL
+    tmb_map$logit_p_mix <- NULL
+  }
+  if (!is.null(family$delta)) {
```
Comment (seananderson) [inline R/fit.R:1113]: !is.null followed by == TRUE is more simply `isTRUE(family$delta)`, but that's already been checked above and assigned to `delta`, so lines 1113 and 1114 can be simplified to `if (delta) {`

**Comment by ericward-noaa** [comment]
Forgot that 'delta' is created -- but yes, that simplifies things. Changed:

https://github.com/pbs-assess/sdmTMB/commit/1fe942474ab07bc8343d83e83ce8973c261ba517

**Comment by ericward-noaa** [comment]
Meant to also add this one in, but delta_beta is https://github.com/pbs-assess/sdmTMB/commit/37d4245c76b9f94dd4f531ff099ce9005267e4cf

**Comment by seananderson** [comment]
I think this is look good. I added some more extensive tests of the delta-beta.

Can the following be added?

- [ ] some brief description in the families help docs about the 'mix' families
- [ ] the equations in the 'model description' vignette (which we can pull out into the guide/manual later)
- [ ] basic simulate() tests for the 'mix' families
- [ ] basic predict() tests for the 'mix' families
- [ ] is there any way we can verify the results are precisely correct? Is VAST our only option? If so, we'll have to do something outside the unit tests. I guess it's that or simulating with large sample sizes for many iterations.

**Comment by ericward-noaa** [comment]
Just committed some things that I think adds to all of these. Description in 'families' is minimal for most families -- but could add more. 

On the testing, we can't test this against VAST because I don't think these distributions are implemented. They're also slightly different from how [we've coded them in the past](https://github.com/pfmc-assessments/nwfscDeltaGLM/blob/master/R/fitDeltaGLM.R). Historically, we've estimated separate variances for each component -- here, there's a single scaling parameter (equivalent to CV). 

I think the best we can do is simulating datasets with known parameters (I've included examples in 'tests/test-mix.R' that demonstrate these can be recovered)

**Pull request merged**